proposal/model-registry.md

```diff
@@ -0,0 +1,53 @@
+---
+title: Model Registry
+---
+
+**Introduction**
+
+Model Registry in an ML DevOps environment is a central repository for storing metadata about ML models.
+
+During the development and deployment life cycle of an ML model, various artifacts and metadata are created or associated with the ML model.
+For example, a *Data Scientist* may create training data, parameters, metrics during development. During deployment, an *MLOps Engineer* may work with artifacts such as versioned model data. Model serving will generate metadata about hyper parameters and metrics from inference. 
```

Comment:
Definition from a quick web search: *Hyperparameters* are parameters whose values control the learning process and determine the values of model parameters that a learning algorithm ends up learning. The prefix ‘hyper_’ suggests that they are ‘top-level’ parameters that control the learning process and the model parameters that result from it.

Reply:
OK, so it is a general AI term.

I'm not expert, but reading a little bit more on the internet, I find weird the statement telling that "_Model serving will generate metadata about hyper parameters_". Hyper-parameters doesn't seem to be a concern for serving phase, but more for model tuning.

I'm OK with metrics, assuming enhancements are needed because AFAIK, we just expose metrics for consumption.

Reply:
The Model Registry itself is going to be agnostic as to the semantics or nature and purpose of metadata being stored. 
As long as a shape/schema of metadata is provided to it, it's going to store it in a DB and support retrieving it. So whether they are hyper-parameters or metrics, they are doing to be stored in the repository as some metadata type and it's property values. 
How the metadata is interpreted and used is also up to the client retrieving the metadata.

Reply:
OK, I can understand the Model Registry being metadata-agnostic and I'm fine with it.

Yet, the example about _Model Serving generating metadata_ is still misleading to me because, AFAIK, we generating metadata is not in the roadmap for _model serving_.

Reply:
I'm with you now. The proposal talks about the general idea that model serving like other ODH components might have metadata that it could expose through the model registry. 
It's not a requirement at the outset, but it could be something it could do in the future. Does that make sense?

Reply:
To elaborate, consider the use case where a user downloads a model off of an external repository like huggingface. There is no data scientist training experiment and run to record in the model registry in the beginning. But, is it possible that model serving could introspect the imported model and determine it's hyperparameters and log whats being used for the version being served?

Reply:
> It's not a requirement at the outset, but it could be something it could do in the future. Does that make sense?

Yes, it does make sense that enhancements on the other components to interact with the model registry may be planned for the future.

> is it possible that model serving could introspect the imported model and determine it's hyperparameters and log whats being used for the version being served?

I do not think this is possible.

---

Re-thinking about this, these are just examples of use-cases. I'm fully :+1: on giving examples, as that helps to realize the value of the model registry.

Perhaps, what is noisy to me is that the hyperparameters example talks about something that cannot be done _at the moment_ and implies that some other enhancements are needed. I'd be more comfortable if the given examples leave aside any enhancement suggestions. If we want to talk about hyperparameters, I'd change the example to something like "the Data Scientist may want to record in the model registry some metadata about the hyperparameters used to train the model", which would be a more realistic example, given current state of our tooling.

So, see this as a nitpick. You don't need to change anything in the proposal. Although if you do, you remove the nit :slightly_smiling_face: .

Reply:
I'll reword it to remove the suggestion that it *has* to be something model serving does. Perhaps the hyperparameters metadata could also come from the model import process while being registered in the model registry.